File: geeksforgeeks/w4-p3-reverse-a-doubly-linked-list/Solution.c
/*
struct Node {
    int data;
    struct Node *next;
    struct Node *prev;
};
*/
struct Node* reverse(struct Node* head) {
    struct Node* temp = NULL;
    struct Node* curr = head;

    while (curr != NULL) {
        // swap next and prev
        temp = curr->prev;
        curr->prev = curr->next;
        curr->next = temp;
        // move to next node (curr->prev)
        curr = curr->prev;
    }

    // adjust head to the new front node
    if (temp != NULL)
        head = temp->prev;

    return head;
}

File: geeksforgeeks/w4-p3-reverse-a-doubly-linked-list/Solution.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None
        self.prev = None

class Solution:
    def reverse(self, head):
        if not head:
            return None

        curr = head
        prev = None
        
        while curr:
            # Swap next and prev pointers
            curr.prev, curr.next = curr.next, curr.prev
            # Move prev to current before going to next node
            prev = curr
            # Move to next node (curr.prev)
            curr = curr.prev
        
        # prev will be the new head after reversal
        return prev
